File: 36-valid-sudoku/valid-sudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function (board) {
    let cols = new Map();
    let rows = new Map();
    let squares = new Map();

    for (let i = 0; i < 9; i++) {
        for (let j = 0; j < 9; j++) {
            if (board[i][j] == '.') {
                continue;
            }

            let rowSet = rows.get(i) || new Set();
            let colSet = cols.get(j) || new Set();
            let squareSet = squares.get(Math.floor(i / 3) + ',' + Math.floor(j / 3)) || new Set();

            if (rowSet.has(board[i][j]) || colSet.has(board[i][j]) || squareSet.has(board[i][j])) {
                return false;
            }

            rowSet.add(board[i][j]);
            colSet.add(board[i][j]);
            squareSet.add(board[i][j]);

            rows.set(i, rowSet);
            cols.set(j, colSet);
            squares.set(Math.floor(i / 3) + ',' + Math.floor(j / 3), squareSet);
        }
    }
    return true;
};

File: 36-valid-sudoku/valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        Set<Character>[] rows = new HashSet[9];
        Set<Character>[] cols = new HashSet[9];
        Set<Character>[] boxes = new HashSet[9];

        // Step 1: Initializing Sets
        for (int i = 0; i < 9; i++) {
            rows[i] = new HashSet<>();
            cols[i] = new HashSet<>();
            boxes[i] = new HashSet<>();
        }

        // Step 2: Iterating Over Board
        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {
                char ch = board[r][c];

                if (ch == '.') continue;

                int boxIndex = (r / 3) * 3 + (c / 3);

                if (rows[r].contains(ch) || cols[c].contains(ch) || boxes[boxIndex].contains(ch)) {
                    return false;
                }

                rows[r].add(ch);
                cols[c].add(ch);
                boxes[boxIndex].add(ch);
            }
        }

        return true;
    }
}